sys/net/application_layer/gcoap/fileserver.c

```diff
@@ -0,0 +1,369 @@
+/*
+ * Copyright (C) 2020 chrysn
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @{
+ *
+ * @file
+ * @brief       CoAP file server implementation
+ *
+ * @author      Christian Amsüss <chrysn@fsfe.org>
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "checksum/fletcher32.h"
+#include "net/gcoap/fileserver.h"
+#include "net/gcoap.h"
+#include "vfs.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/** Maximum length of an expressible path, including the trailing 0 character. */
+#define COAPFILESERVER_PATH_MAX (64)
+
+/** Data extracted from a request on a file */
+struct requestdata {
+    /** 0-terminated expanded file name in the VFS */
+    char namebuf[COAPFILESERVER_PATH_MAX];
+    uint32_t blocknum2;
+    unsigned int szx2; /* would prefer uint8_t but that's what coap_get_blockopt gives */
+    uint32_t etag;
+    bool etag_sent;
+};
+
+/**
+ * @brief  Return true if path/name is a directory.
+ */
+static bool entry_is_dir(char *path, const char *name)
+{
+    struct stat stat;
+    size_t path_len = strlen(path);
+    size_t name_len = strlen(name);
+
+    if (path_len + name_len + 1 >= COAPFILESERVER_PATH_MAX) {
+        return false;
+    }
+
+    /* re-use path buffer, it is already COAPFILESERVER_PATH_MAX bytes long */
+    path[path_len] = '/';
+    path[path_len + 1] = 0;
+    strncat(path, name, COAPFILESERVER_PATH_MAX - 1);
+
+    if (vfs_stat(path, &stat)) {
+        DEBUG("vfs_stat(%s) failed\n", path);
+    }
+
+    path[path_len] = 0;
+    return (stat.st_mode & S_IFMT) == S_IFDIR;
+}
+
+static unsigned _count_char(const char *s, char c)
+{
+    unsigned count = 0;
+    while (*s) {
+        if (*s == c) {
+            ++count;
+        }
+        ++s;
+    }
+
+    return count;
+}
+
+/** Build an ETag based on the given file's VFS stat. If the stat fails,
+ * returns the error and leaves etag in any state; otherwise there's an etag
+ * in the stattag's field */
+static int stat_etag(const char *filename, uint32_t *etag)
+{
+    struct stat stat;
+    int err = vfs_stat(filename, &stat);
+    if (err < 0) {
+        return err;
+    }
+
+    /* Normalizing fields whose value can change without affecting the ETag */
+    stat.st_nlink = 0;
+    memset(&stat.st_atime, 0, sizeof(stat.st_atime));
```

Comment:
`cpu/avr8_common/avr_libc_extra/include/vendor/sys/stat.h` does not `#define st_atime st_atim.tv_sec` and we should better `memset()` the whole `struct timespec` 

```suggestion
    memset(&stat.st_atim, 0, sizeof(stat.st_atim));
```

Reply:
uh now esp* and pic32 don't know `.st_atim` but only `.st_atime` :confused: